sw/device/tests/entropy_src_csrng_test.c

```diff
@@ -0,0 +1,223 @@
+#include "sw/device/lib/dif/dif_csrng.h"
+#include "sw/device/lib/dif/dif_edn.h"
+#include "sw/device/lib/dif/dif_entropy_src.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/csrng_testutils.h"
+#include "sw/device/lib/testing/entropy_testutils.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_csrng_t csrng;
+static dif_edn_t edn0;
+static dif_edn_t edn1;
+static dif_entropy_src_t entropy_src;
+
+static dif_rv_plic_t plic;
+static volatile top_earlgrey_plic_peripheral_t plic_peripheral;
+static volatile dif_rv_plic_irq_id_t irq_id;
+static volatile dif_csrng_irq_t irq;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+enum {
+  /**
+   * The size of the buffer used in firmware to process the entropy bits in
+   * firmware override mode.
+   */
+  kTestParamFifoBufferSize = 12,
+  /**
+   * The number of test iterations per target.
+   */
+  kTestParamNumIterationsSim = 1,
+  kTestParamNumIterationsOther = 100,
+};
+
+/**
+ * Initializes the peripherals used in this test.
+ */
+static void init_peripherals(void) {
+  CHECK_DIF_OK(dif_csrng_init(
+      mmio_region_from_addr(TOP_EARLGREY_CSRNG_BASE_ADDR), &csrng));
+  CHECK_DIF_OK(
+      dif_edn_init(mmio_region_from_addr(TOP_EARLGREY_EDN0_BASE_ADDR), &edn0));
+  CHECK_DIF_OK(
+      dif_edn_init(mmio_region_from_addr(TOP_EARLGREY_EDN1_BASE_ADDR), &edn1));
+  CHECK_DIF_OK(dif_entropy_src_init(
+      mmio_region_from_addr(TOP_EARLGREY_ENTROPY_SRC_BASE_ADDR), &entropy_src));
+  CHECK_DIF_OK(dif_rv_plic_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR), &plic));
+}
+
+/**
+ * Enables the CSRNG entropy request interrupt.
+ */
+static void csrng_entropy_req_irq_enable(void) {
+  irq_external_ctrl(false);
+  irq_global_ctrl(false);
+
+  // Clear interrupt context variables. These will be set by the ISR routine
+  // `ottf_external_isr()`.
+  irq = UINT32_MAX;
+  irq_id = UINT32_MAX;
+  plic_peripheral = UINT32_MAX;
+
+  dif_rv_plic_irq_id_t irq_id = kTopEarlgreyPlicIrqIdCsrngCsEntropyReq;
+  CHECK_DIF_OK(dif_rv_plic_irq_set_priority(&plic, irq_id, /*priority=*/1u));
+
+  CHECK_DIF_OK(dif_rv_plic_irq_set_enabled(
+      &plic, irq_id, kTopEarlgreyPlicTargetIbex0, kDifToggleEnabled));
+  CHECK_DIF_OK(dif_rv_plic_target_set_threshold(
+      &plic, kTopEarlgreyPlicTargetIbex0, /*threshold=*/0u));
+
+  CHECK_DIF_OK(dif_csrng_irq_set_enabled(&csrng, kDifCsrngIrqCsEntropyReq,
+                                         kDifToggleEnabled));
+
+  irq_global_ctrl(true);
+  irq_external_ctrl(true);
+}
+
+/**
+ * Blocks until a CSRNG entropy request interrupt is received.
+ */
+static void csrng_entropy_req_irq_block_wait(void) {
+  // The interrupt can come before we enter sleep, so we check beforehand.
+  while (true) {
+    irq_global_ctrl(false);
```

Comment:
this is a little bit confusing to me. Does `irq_global_ctrl(false)` not disable the interrupt completely? or does it just disabling jumping into the ISR routine?

Reply:
It seems `wfi` is able to return on an interrupt even if interrupts are globally disabled, so I guess this means this only disables jumping into the ISR routine. I am not sure if this is the expected behavior though. I refactored the code to simplify this.